src/field_theory/perfect_closure.lean

```diff
@@ -18,65 +18,82 @@ open function
 
 section defs
 
-variables (K : Type u) [field K] (p : ℕ) [fact p.prime] [char_p K p]
+variables (R : Type u) [comm_semiring R] (p : ℕ) [fact p.prime] [char_p R p]
 
-/-- A perfect field is a field of characteristic p that has p-th root. -/
-class perfect_field (K : Type u) [field K] (p : ℕ) [fact p.prime] [char_p K p] : Type u :=
-(pth_root' : K → K)
-(frobenius_pth_root' : ∀ x, frobenius K p (pth_root' x) = x)
+/-- A perfect ring is a ring of characteristic p that has p-th root. -/
+class perfect_ring : Type u :=
+(pth_root' : R → R)
+(frobenius_pth_root' : ∀ x, frobenius R p (pth_root' x) = x)
+(pth_root_frobenius' : ∀ x, pth_root' (frobenius R p x) = x)
 
-/-- Frobenius automorphism of a perfect field. -/
-def frobenius_equiv [perfect_field K p] : K ≃+* K :=
-{ inv_fun := perfect_field.pth_root' p,
-  left_inv := λ x, frobenius_inj K p $ perfect_field.frobenius_pth_root' _,
-  right_inv := perfect_field.frobenius_pth_root',
-  .. frobenius K p }
+/-- Frobenius automorphism of a perfect ring. -/
+def frobenius_equiv [perfect_ring R p] : R ≃+* R :=
+{ inv_fun := perfect_ring.pth_root' p,
+  left_inv := perfect_ring.pth_root_frobenius',
+  right_inv := perfect_ring.frobenius_pth_root',
+  .. frobenius R p }
 
-/-- `p`-th root of a number in a `perfect_field` as a `ring_hom`. -/
-def pth_root [perfect_field K p] : K →+* K :=
-(frobenius_equiv K p).symm.to_ring_hom
+/-- `p`-th root of an element in a `perfect_ring` as a `ring_hom`. -/
+def pth_root [perfect_ring R p] : R →+* R :=
+(frobenius_equiv R p).symm
 
 end defs
 
 section
 
-variables {K : Type u} [field K] {L : Type v} [field L] (f : K →* L) (g : K →+* L)
-  {p : ℕ} [fact p.prime] [char_p K p] [perfect_field K p] [char_p L p] [perfect_field L p]
+variables {R : Type u} [comm_semiring R] {S : Type v} [comm_semiring S] (f : R →* S) (g : R →+* S)
+  {p : ℕ} [fact p.prime] [char_p R p] [perfect_ring R p] [char_p S p] [perfect_ring S p]
 
-@[simp] lemma coe_frobenius_equiv : ⇑(frobenius_equiv K p) = frobenius K p := rfl
+@[simp] lemma coe_frobenius_equiv : ⇑(frobenius_equiv R p) = frobenius R p := rfl
 
-@[simp] lemma  coe_frobenius_equiv_symm : ⇑(frobenius_equiv K p).symm = pth_root K p := rfl
+@[simp] lemma coe_frobenius_equiv_symm : ⇑(frobenius_equiv R p).symm = pth_root R p := rfl
 
-@[simp] theorem frobenius_pth_root (x : K) : frobenius K p (pth_root K p x) = x :=
-(frobenius_equiv K p).apply_symm_apply x
+@[simp] theorem frobenius_pth_root (x : R) : frobenius R p (pth_root R p x) = x :=
+(frobenius_equiv R p).apply_symm_apply x
 
-@[simp] theorem pth_root_frobenius (x : K) : pth_root K p (frobenius K p x) = x :=
-(frobenius_equiv K p).symm_apply_apply x
+@[simp] theorem pth_root_pow_p (x : R) : pth_root R p x ^ p = x :=
```

Comment:
Please mention the name change in the top post of the PR. That way it ends up in the git log where people can easily find it.

Reply:
This is not a name change (see L57).